cli/internal/taskhash/taskhash.go

```diff
@@ -129,45 +99,51 @@ func (th *Tracker) CalculateFileHashes(
 			return fmt.Errorf("missing pipeline entry %v", taskID)
 		}
 
-		pfs := &packageFileSpec{
-			pkg:    pkgName,
-			inputs: taskDefinition.Inputs,
+		pfs := &packageFileHashInputs{
+			taskID,
+			taskDefinition,
+			packageName,
 		}
 
 		hashTasks.Add(pfs)
 	}
 
-	hashes := make(map[packageFileHashKey]string, len(hashTasks))
-	hashObjects := make(map[packageFileHashKey]map[turbopath.AnchoredUnixPath]string, len(hashTasks))
-	hashQueue := make(chan *packageFileSpec, workerCount)
+	hashes := make(map[string]string, len(hashTasks))
+	hashObjects := make(map[string]map[turbopath.AnchoredUnixPath]string, len(hashTasks))
+	hashQueue := make(chan *packageFileHashInputs, workerCount)
 	hashErrs := &errgroup.Group{}
 
 	for i := 0; i < workerCount; i++ {
 		hashErrs.Go(func() error {
-			for packageFileSpec := range hashQueue {
-				pkg, ok := workspaceInfos.PackageJSONs[packageFileSpec.pkg]
+			for packageFileHashInputs := range hashQueue {
+				pkg, ok := workspaceInfos.PackageJSONs[packageFileHashInputs.packageName]
 				if !ok {
-					return fmt.Errorf("cannot find package %v", packageFileSpec.pkg)
+					return fmt.Errorf("cannot find package %v", packageFileHashInputs.packageName)
 				}
-				hashObject, err := packageFileSpec.getHashObject(pkg, repoRoot)
+
+				// Get the hashes of each file, keyed by the path.
+				hashObject, err := hashing.GetPackageFileHashes(repoRoot, pkg.Dir, packageFileHashInputs.taskDefinition.Inputs)
 				if err != nil {
 					return err
 				}
-				hash, err := packageFileSpec.hash(hashObject)
+
+				// Get the combined hash of all the files.
+				hash, err := fs.HashObject(hashObject)
 				if err != nil {
 					return err
 				}
+
+				// Save off the hash information, keyed by package task.
 				th.mu.Lock()
-				pfsKey := packageFileSpec.ToKey()
-				hashes[pfsKey] = hash
-				hashObjects[pfsKey] = hashObject
+				hashes[packageFileHashInputs.taskID] = hash
+				hashObjects[packageFileHashInputs.taskID] = hashObject
```

Comment:
`taskID` is a _much_ better key, making the whole thing understandable.

Reply:
yesss. this tripped me up a lot when i was working on this code, thank you

Reply:
I think this is going to result in rehashing the same package multiple times if there is more than one task in that package.

For instance, if `test` depends on `build`, and neither specifies `inputs`, this will double the amount of hashing that we do.

Reply:
Ok, I doubled checked what I wrote and it looks like there's a bug where it's not doing what we expect. This change _would_ result in rehashing the same package multiple times if we were already properly deduping sets of files to hash, but we are not. 

The reason we are not currently properly deduping is because we are adding the _addresses_ of `packageFileSpec` instances to a set, which are all going to be unique, even if the actual values are equivalent. 

The original author (me) probably should've noticed this given the effort to produce a key for a map later on, and knowing that the set is implemented via a map under the hood. I'm going to file an issue to fix this post-port, since it will be much easier to do the struct value equality in Rust.

Reply:
The actual deduping I want to do is to push all files into a set, do a set intersection, and only hash files that don't already have a hash. There's no particular reason to tie it to a package or task.